File: packages/rabbitmq/src/producer.ts
/**
 * RabbitMQ Producer Implementation
 */

import amqp, { type ChannelModel, type ConfirmChannel } from 'amqplib';
import {
  BaseProducer,
  type PublishOptions,
  type HealthStatus,
  ConnectionError,
  PublishError,
  generateUUID,
} from '@anyq/core';
import type { RabbitMQProducerConfig } from './config.js';
import { RABBITMQ_DEFAULTS } from './config.js';

/**
 * RabbitMQ Producer
 *
 * Publishes messages to RabbitMQ exchanges
 */
export class RabbitMQProducer<T = unknown> extends BaseProducer<T> {
  private connection: ChannelModel | null = null;
  private channel: ConfirmChannel | null = null;
  private readonly rabbitConfig: RabbitMQProducerConfig;

  constructor(config: RabbitMQProducerConfig) {
    super(config);
    this.rabbitConfig = {
      ...config,
      connection: {
        ...RABBITMQ_DEFAULTS.connection,
        ...config.connection,
      },
      exchange: {
        ...RABBITMQ_DEFAULTS.exchange,
        ...config.exchange,
      },
    };
  }

  /**
   * Connect to RabbitMQ
   */
  async connect(): Promise<void> {
    if (this._connected) {
      return;
    }

    try {
      const { connection: connConfig, exchange: exchangeConfig } = this.rabbitConfig;

      // Build connection URL
      let url = connConfig.url;
      if (connConfig.vhost) {
        url = `${url}/${encodeURIComponent(connConfig.vhost)}`;
      }

      // Connect
      this.connection = await amqp.connect(url, {
        heartbeat: connConfig.heartbeat,
        timeout: connConfig.connectionTimeout,
        channelMax: connConfig.channelMax,
        frameMax: connConfig.frameMax,
      });

      // Create confirm channel for reliable publishing
      this.channel = await this.connection.createConfirmChannel();

      // Assert exchange
      await this.channel.assertExchange(exchangeConfig.name, exchangeConfig.type, {
        durable: exchangeConfig.durable,
        autoDelete: exchangeConfig.autoDelete,
        internal: exchangeConfig.internal,
        arguments: exchangeConfig.arguments,
      });

      // Handle connection errors
      this.connection.on('error', (err) => {
        this.logger.error('Connection error', { error: err.message });
        this._connected = false;
      });

      this.connection.on('close', () => {
        this.logger.warn('Connection closed');
        this._connected = false;
      });

      this._connected = true;
      this.logger.info('RabbitMQ producer connected', {
        url: connConfig.url,
        exchange: exchangeConfig.name,
        type: exchangeConfig.type,
      });
    } catch (error) {
      throw new ConnectionError(
        'Failed to connect to RabbitMQ',
        error instanceof Error ? error : undefined
      );
    }
  }

  /**
   * Disconnect from RabbitMQ
   */
  async disconnect(): Promise<void> {
    try {
      if (this.channel) {
        await this.channel.close();
        this.channel = null;
      }

      if (this.connection) {
        await this.connection.close();
        this.connection = null;
      }

      this._connected = false;
      this.logger.info('RabbitMQ producer disconnected');
    } catch (error) {
      this.logger.error('Error during disconnect', { error });
    }
  }

  /**
   * Publish a single message
   */
  async publish(body: T, options?: PublishOptions): Promise<string> {
    if (!this.channel) {
      throw new ConnectionError('Producer not connected');
    }

    const messageId = generateUUID();
    const routingKey = options?.key ?? this.rabbitConfig.routingKey ?? '';
    const exchange = this.rabbitConfig.exchange.name;

    try {
      const serializedBody = this.serializer.serialize(body);
      const content = Buffer.isBuffer(serializedBody)
        ? serializedBody
        : Buffer.from(serializedBody);

      // Build message properties
      const messageOptions: amqp.Options.Publish = {
        messageId,
        timestamp: Date.now(),
        contentType: 'application/json',
        deliveryMode: this.rabbitConfig.persistent !== false ? 2 : 1, // 2 = persistent
        mandatory: this.rabbitConfig.mandatory,
        priority: options?.priority,
        correlationId: options?.correlationId,
        replyTo: options?.replyTo,
        expiration: options?.ttlMs?.toString(),
        headers: options?.headers as Record<string, unknown>,
      };

      // Publish with confirmation
      return new Promise<string>((resolve, reject) => {
        const published = this.channel!.publish(
          exchange,
          routingKey,
          content,
          messageOptions,
          (err) => {
            if (err) {
              reject(new PublishError('Publish failed', { cause: err }));
            } else {
              this.logger.debug('Message published', {
                messageId,
                exchange,
                routingKey,
              });
              resolve(messageId);
            }
          }
        );

        if (!published) {
          // Channel buffer is full
          this.channel!.once('drain', () => {
            this.logger.warn('Channel drained after backpressure');
          });
        }
      });
    } catch (error) {
      throw new PublishError('Failed to publish message', {
        cause: error instanceof Error ? error : undefined,
      });
    }
  }

  /**
   * Publish multiple messages
   */
  async publishBatch(
    messages: Array<{ body: T; options?: PublishOptions }>
  ): Promise<string[]> {
    const results: string[] = [];

    for (const { body, options } of messages) {
      const messageId = await this.publish(body, options);
      results.push(messageId);
    }

    // Wait for all confirms
    await this.channel?.waitForConfirms();

    this.logger.debug('Batch published', {
      count: results.length,
      exchange: this.rabbitConfig.exchange.name,
    });

    return results;
  }

  /**
   * Flush pending messages
   */
  async flush(): Promise<void> {
    if (this.channel) {
      await this.channel.waitForConfirms();
    }
  }

  /**
   * Get health status
   */
  async healthCheck(): Promise<HealthStatus> {
    const start = Date.now();

    try {
      if (!this._connected || !this.channel) {
        return {
          healthy: false,
          connected: false,
          error: 'Not connected',
        };
      }

      // Check exchange exists
      await this.channel.checkExchange(this.rabbitConfig.exchange.name);

      return {
        healthy: true,
        connected: true,
        latencyMs: Date.now() - start,
        details: {
          exchange: this.rabbitConfig.exchange.name,
          type: this.rabbitConfig.exchange.type,
        },
      };
    } catch (error) {
      return {
        healthy: false,
        connected: false,
        latencyMs: Date.now() - start,
        error: error instanceof Error ? error.message : String(error),
      };
    }
  }
}

/**
 * Create a RabbitMQ producer instance
 */
export function createRabbitMQProducer<T = unknown>(
  config: RabbitMQProducerConfig
): RabbitMQProducer<T> {
  return new RabbitMQProducer<T>(config);
}

File: go/rabbitmq/producer.go
package rabbitmq

import (
	"context"
	"time"

	amqp "github.com/rabbitmq/amqp091-go"

	"github.com/sns45/anyq/go/core"
)

// Producer publishes messages to a RabbitMQ exchange. It mirrors the TypeScript
// RabbitMQProducer: it asserts the configured exchange on connect and publishes
// with publisher confirms when ConfirmMode is enabled.
type Producer struct {
	core.BaseProducer
	cfg ProducerConfig

	conn    *amqp.Connection
	channel *amqp.Channel
}

// NewProducer builds a RabbitMQ producer.
func NewProducer(cfg ProducerConfig) *Producer {
	cfg.applyDefaults()
	p := &Producer{cfg: cfg}
	p.InitBase(cfg.BaseQueueConfig)
	return p
}

// Connect dials the broker, opens a channel, optionally enables confirms, and
// asserts the exchange.
func (p *Producer) Connect(ctx context.Context) error {
	if p.IsConnected() {
		return nil
	}

	conn, err := amqp.DialConfig(dialURL(p.cfg.Connection), amqp.Config{
		Heartbeat:  p.cfg.Connection.Heartbeat,
		Dial:       amqp.DefaultDial(p.cfg.Connection.ConnectionTimeout),
		ChannelMax: uint16(p.cfg.Connection.ChannelMax),
		FrameSize:  p.cfg.Connection.FrameMax,
	})
	if err != nil {
		return core.NewConnectionError("failed to connect to RabbitMQ", err)
	}

	ch, err := conn.Channel()
	if err != nil {
		_ = conn.Close()
		return core.NewConnectionError("failed to open RabbitMQ channel", err)
	}

	if p.cfg.ConfirmMode {
		if err := ch.Confirm(false); err != nil {
			_ = ch.Close()
			_ = conn.Close()
			return core.NewConnectionError("failed to enable publisher confirms", err)
		}
	}

	if p.cfg.Exchange.Name != "" {
		if err := ch.ExchangeDeclare(
			p.cfg.Exchange.Name,
			p.cfg.Exchange.Type,
			p.cfg.Exchange.Durable,
			p.cfg.Exchange.AutoDelete,
			p.cfg.Exchange.Internal,
			false,
			amqp.Table(p.cfg.Exchange.Arguments),
		); err != nil {
			_ = ch.Close()
			_ = conn.Close()
			return core.NewConnectionError("failed to declare exchange", err)
		}
	}

	p.conn = conn
	p.channel = ch
	p.SetConnected(true)
	p.Logger.Info("RabbitMQ producer connected", map[string]any{
		"exchange": p.cfg.Exchange.Name,
		"type":     p.cfg.Exchange.Type,
	})
	return nil
}

// Disconnect closes the channel and connection.
func (p *Producer) Disconnect(ctx context.Context) error {
	if !p.IsConnected() {
		return nil
	}
	if p.channel != nil {
		_ = p.channel.Close()
		p.channel = nil
	}
	if p.conn != nil {
		_ = p.conn.Close()
		p.conn = nil
	}
	p.SetConnected(false)
	p.Logger.Info("RabbitMQ producer disconnected", nil)
	return nil
}

// Publish sends a single message. The routing key is options.Key, falling back
// to the configured RoutingKey. When ConfirmMode is enabled the publish waits
// for a broker confirm bounded by ConfirmTimeout.
func (p *Producer) Publish(ctx context.Context, body []byte, opts *core.PublishOptions) (string, error) {
	if p.channel == nil {
		return "", core.NewConnectionError("producer not connected", nil)
	}

	messageID := core.GenerateMessageID()
	routingKey := p.cfg.RoutingKey
	var headers core.MessageHeaders
	var priority uint8
	var correlationID, replyTo, expiration string
	if opts != nil {
		if opts.Key != "" {
			routingKey = opts.Key
		}
		headers = opts.Headers
		if opts.Priority != nil {
			priority = *opts.Priority
		}
		correlationID = opts.CorrelationID
		replyTo = opts.ReplyTo
		if opts.TTL > 0 {
			expiration = itoa(int(opts.TTL / time.Millisecond))
		}
	}

	deliveryMode := uint8(amqp.Persistent)
	if !p.cfg.Persistent {
		deliveryMode = amqp.Transient
	}

	pub := amqp.Publishing{
		Headers:       toTable(headers),
		ContentType:   "application/json",
		DeliveryMode:  deliveryMode,
		Priority:      priority,
		CorrelationId: correlationID,
		ReplyTo:       replyTo,
		Expiration:    expiration,
		MessageId:     messageID,
		Timestamp:     time.Now(),
		Body:          body,
	}

	if p.cfg.ConfirmMode {
		conf, err := p.channel.PublishWithDeferredConfirmWithContext(
			ctx, p.cfg.Exchange.Name, routingKey, p.cfg.Mandatory, false, pub)
		if err != nil {
			return "", core.NewPublishError("failed to publish message", true, err)
		}
		confirmCtx := ctx
		if p.cfg.ConfirmTimeout > 0 {
			var cancel context.CancelFunc
			confirmCtx, cancel = context.WithTimeout(ctx, p.cfg.ConfirmTimeout)
			defer cancel()
		}
		acked, err := conf.WaitContext(confirmCtx)
		if err != nil {
			return "", core.NewPublishError("publish confirm wait failed", true, err)
		}
		if !acked {
			return "", core.NewPublishError("publish nacked by broker", true, nil)
		}
	} else {
		if err := p.channel.PublishWithContext(
			ctx, p.cfg.Exchange.Name, routingKey, p.cfg.Mandatory, false, pub); err != nil {
			return "", core.NewPublishError("failed to publish message", true, err)
		}
	}

	p.Logger.Debug("message published", map[string]any{
		"messageId": messageID, "exchange": p.cfg.Exchange.Name, "routingKey": routingKey,
	})
	return messageID, nil
}

// PublishBatch publishes each item in order and returns the assigned ids.
func (p *Producer) PublishBatch(ctx context.Context, items []core.BatchItem) ([]string, error) {
	if p.channel == nil {
		return nil, core.NewConnectionError("producer not connected", nil)
	}
	ids := make([]string, 0, len(items))
	for _, item := range items {
		id, err := p.Publish(ctx, item.Body, item.Options)
		if err != nil {
			return ids, err
		}
		ids = append(ids, id)
	}
	p.Logger.Debug("batch published", map[string]any{"count": len(ids), "exchange": p.cfg.Exchange.Name})
	return ids, nil
}

// Flush is a no-op: each Publish already waits for its confirm when ConfirmMode
// is enabled.
func (p *Producer) Flush(ctx context.Context) error { return nil }

// HealthCheck reports producer health.
func (p *Producer) HealthCheck(ctx context.Context) (core.HealthStatus, error) {
	start := time.Now()
	if !p.IsConnected() || p.channel == nil {
		return core.HealthStatus{Healthy: false, Connected: false, Error: "Not connected"}, nil
	}
	return core.HealthStatus{
		Healthy:   true,
		Connected: true,
		LatencyMs: time.Since(start).Milliseconds(),
		Details:   map[string]any{"exchange": p.cfg.Exchange.Name, "type": p.cfg.Exchange.Type},
	}, nil
}

// toTable converts core MessageHeaders ([]byte values) into an amqp.Table.
func toTable(h core.MessageHeaders) amqp.Table {
	if len(h) == 0 {
		return nil
	}
	t := make(amqp.Table, len(h))
	for k, v := range h {
		t[k] = string(v)
	}
	return t
}

var _ core.Producer = (*Producer)(nil)
